File: Src/node/js/health_monitoring.js
/* eslint-disable no-undef */
let cpuData;
let memData;
let diskData;

let cpuOptions;
let memOptions;
let diskOptions;

let cpuChart;
let memChart;
let diskChart;

let index = 0;
let first = true;

// load current chart package
google.charts.load('current', {
  packages: ['corechart', 'line'],
});

// a chart which shows the percentage usage of the cpu of the corresponding device
function drawCPUChart() {
  // create data object with default value
  cpuData = google.visualization.arrayToDataTable([
    ['Year', 'Usage'],
    [0, 0.0],
  ]);

  // create options object with titles, colors, etc.
  cpuOptions = {
    title: 'CPU',
    hAxis: {
      title: 'Time',
    },
    vAxis: {
      title: '',
    },
    trendlines: {
      0: {
        color: 'purple',
        labelInLegend: 'trend',
        visibleInLegend: true,
        opacity: 0.4,
      },
    },
  };

  // draw chart on load
  cpuChart = new google.visualization.LineChart(document.getElementById('cpu_chart_div'));
  cpuChart.draw(cpuData, cpuOptions);
}

// a chart which shows the usage and total memory of the disk of the corresponding device
function drawMEMChart() {
  memData = google.visualization.arrayToDataTable([
    ['Year', 'Usage'], // 'Capacity'],
    [0, 0.0], // 0.0],
  ]);

  memOptions = {
    title: 'RAM',
    hAxis: {
      title: 'Time',
    },
    vAxis: {
      title: '',
    },
    trendlines: {
      0: {
        color: 'blue',
        labelInLegend: 'trend',
        visibleInLegend: true,
        opacity: 0.4,
      },
    },
  };
  // draw chart on load
  memChart = new google.visualization.LineChart(document.getElementById('mem_chart_div'));
  memChart.draw(memData, memOptions);
}

// a chart which shows the read and write of the disk of the corresponding device
function drawDISKChart() {
  diskData = google.visualization.arrayToDataTable([
    ['Year', 'Read', 'Write'],
    [0, 0.0, 0.0],
  ]);

  diskOptions = {
    title: 'DISK',
    hAxis: {
      title: 'Time',
    },
    vAxis: {
      title: '',
    },
    /*trendlines: {
      0: {
        color: 'blue',
        labelInLegend: 'r-trend',
        visibleInLegend: true,
        opacity: 0.4,
      },
      1: {
        color: 'red',
        labelInLegend: 'w-trend',
        visibleInLegend: true,
        opacity: 0.4,
      },
    },*/
  };
  // draw chart on load
  diskChart = new google.visualization.LineChart(document.getElementById('disk_chart_div'));
  diskChart.draw(diskData, diskOptions);
}

// set callback function when api loaded
google.charts.setOnLoadCallback(drawCPUChart);

google.charts.setOnLoadCallback(drawMEMChart);

google.charts.setOnLoadCallback(drawDISKChart);

const clientId = `mqttjs_${Math.random().toString(16).substr(2, 8)}`;

const host = 'ws://localhost:9001/mqtt';

// these are the options which are used for the mqtt connection with the broker
const options = {
  keepalive: 30,
  protocolId: 'MQTT',
  protocolVersion: 4,
  clean: true,
  reconnectPeriod: 1000,
  connectTimeout: 30 * 1000,
  will: {
    topic: 'health_monitoring',
    payload: 'Connection Closed abnormally..!',
    qos: 0,
    retain: false,
  },
  rejectUnauthorized: false,
};

console.log('connecting mqtt client');
const client = mqtt.connect(host, options);

// we add the incoming messages from the health monitoring and put them in our charts
client.on('message', (topic, message, packet) => {
  // console.log('Received Message: ' + message.toString() + '\nOn topic: ' + topic)

  data = JSON.parse(message);

  if (first) {
    cpuOptions.vAxis.title = 'Usage in %';
    memOptions.vAxis.title = `Usage in ${data[1].used_mem[1]}`;
    diskOptions.vAxis.title = `Usage in ${data[2].disk_read[1]}`;
    first = false;
  }

  cpuData.addRow([index, data[0].cpu_usage]);
  cpuChart.draw(cpuData, cpuOptions);
  memData.addRow([index, data[1].used_mem[0]]); //, data[1].total_mem[0]]);
  memChart.draw(memData, memOptions);
  diskData.addRow([index, data[2].disk_read[0], data[2].disk_write[0]]);
  diskChart.draw(diskData, diskOptions);

  index += 1;
});

/* 
    we can change the interval at which messages are sent by health monitoring

    source: https://stackoverflow.com/questions/175739/built-in-way-in-javascript-to-check-if-a-string-is-a-valid-number
    first we check whether the value is a numeric number 
*/
function isNumeric(str) {
  return !str.isNaN && !parseFloat(str.isNaN);
}

// now we can send the value to the health monitoring via mqtt
function changeInterval() {
  value = document.getElementById('interval').value;
  if (isNumeric(value)) {
    client.publish('health_monitoring_interval', value);
  }
}

// generic callback functions which normally will never be called

client.on('error', (err) => {
  console.log('Connection error: ', err);
  client.end();
});

client.on('reconnect', () => {
  console.log('Reconnecting...');
});

client.on('connect', () => {
  console.log(`Client connected:${clientId}`);
  client.subscribe('health_monitoring', { qos: 0 });
});

client.on('close', () => {
  console.log(`${clientId} disconnected`);
});

File: Src/python/health_monitoring.py
from typing import Any, Dict
from client.mqtt_client import mqtt_client

import psutil
import time
import json
import argparse


def parse_arguments() -> Dict[str, Any]:
    parser = argparse.ArgumentParser(description='Description of your program')
    parser.add_argument(
        '-ho', '--host', help='address of host | default: localhost', default='localhost', type=str)
    parser.add_argument(
        '-p', '--port', help='network port | default: 9001', default=9001, type=int)
    parser.add_argument(
        '-ri', '--reconnect_interval', help='connecting to broker | default: 3', default=3, type=float)
    parser.add_argument(
        '-pi', '--publish_interval', help='publishing messages | default: 1', default=1, type=float)
    parser.add_argument(
        '-x', '--publish_x', help='publishing interval min | default: 0.1', default=0.1, type=float)
    parser.add_argument(
        '-y', '--publish_y', help='publishing interval max | default: 100.0', default=100.0, type=float)
    return vars(parser.parse_args())


'''
Scale bytes to its proper format
source: https://stackoverflow.com/questions/1094841/get-human-readable-version-of-file-size
'''


def size_of(num: int, suffix='B'):
    for unit in ['', 'Ki', 'Mi', 'Gi', 'Ti', 'Pi', 'Ei', 'Zi']:
        if abs(num) < 1024.0:
            # return f'{num:3.1f}{unit}{suffix}'
            return float(f'{num:3.1f}'), f'{unit}{suffix}'
        num /= 1024.0
    return f'{num:.1f}Yi{suffix}'


'''
    return percentage usage of total cpu_processors of actual device
'''


def cpu_informations() -> None:
    return {
        'cpu_usage': psutil.cpu_percent(interval=0.025)
    }


'''
    return total memory and current used memory usage of actual device
'''


def memory_informations() -> None:
    mem = psutil.virtual_memory()
    return {
        # 'total_mem': size_of(mem.total),
        'used_mem': size_of(mem.used)
    }


'''
    return total read and write of all disks of actual device
'''


def disk_informations() -> None:
    io_counters = psutil.disk_io_counters()
    return {
        'disk_read': size_of(io_counters.read_count),
        'disk_write': size_of(io_counters.write_count)
    }


'''
    reaction of incoming messages from subscribed topics
    changes the interval in which we publish the messages
'''


def on_message(client, userdata, message) -> None:
    global config
    try:
        new_interval = float(message.payload.decode('utf-8'))
    except ValueError:
        return

    # we just accept values in the following range: [publish_x, publish_y]
    if (new_interval >= config['publish_x'] and new_interval <= config['publish_y']):
        config['publish_interval'] = new_interval
        print('Interval changed to', config['publish_interval'])


if __name__ == '__main__':
    # contains essential configurations for the mqtt client
    global config

    config = parse_arguments()

    client = mqtt_client(config['host'], config['port'],
                         config['reconnect_interval'])

    # subscribe so we can react to interval changes
    client.subcribe('health_monitoring_interval')
    client.set_on_message(on_message)

    # we permanently publish || should be changed in production
    while True:
        cpu = cpu_informations()
        mem = memory_informations()
        disk = disk_informations()
        # dump it in a JSON-object
        stats = json.dumps([cpu, mem, disk])
        # publish our JSON-object to the broker
        client.publish('health_monitoring', stats)
        # wait {publish_interval} seconds
        time.sleep(config['publish_interval'])
